packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts

```diff
@@ -384,23 +388,69 @@ export const useGridSelection = (
 
   const handleCellKeyDown = React.useCallback<GridEventListener<GridEvents.cellKeyDown>>(
     (params, event) => {
+      // Get the most recent cell mode because it may have been changed by another listener
+      if (apiRef.current.getCellMode(params.id, params.field) === GridCellModes.Edit) {
+        return;
+      }
+
+      if (isNavigationKey(event.key) && event.shiftKey) {
+        const focusCell = gridFocusCellSelector(apiRef);
+        if (focusCell && focusCell.id !== params.id) {
+          event.preventDefault();
+
+          const isPreviousRowSelected = apiRef.current.isRowSelected(focusCell.id);
```

Comment:
We're relying on the fact that the focus was not updated yet. If the order of the listeners is changed, or we change to listen for the event in the capture phase, it won't work. Would it work if `lastRowToggled` is used here instead?

Reply:
> We're relying on the fact that the focus was not updated yet.

The opposite, I missed-named this variable, it should be `isNewRowSelected`
I access the previous row through `params.id` and the new one through `focusCell.id`

I don't think we could easily cover all use cases
If you navigate with the keyboard without selecting rows, your "previousRow" still needs to be the cell the focus is currently in, not the last row toggled.
I agree that this is a constraint on the call order and that is why I am putting `useGridKeyboardNavigation` before `useGridSelection`.

Reply:
Sounds fair. I only don't like the constraint on the call order, but I don't have any other simple solution. We could listen to `cellFocusIn` and try to track if the new focus came from the keyboard navigation. It's a way to not depend on the call order but much more complex. We can leave it as it is.

Reply:
Alternatively, the keyboard navigation could publish an event `keyboardNavigation` with the cell before and after.